### PR TITLE
box: Add documentation for `From` impls

### PR DESCRIPTION
**Pull request by daniellimws**
This is a part of #51430. A brief description of the behaviour and examples are added to the documentation.

I am not sure what sort of examples to put for the `From` for `Pin` as my [code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=97c908f44e41c9faeffec5b61d72a03e) doesn't even manage to compile using the nightly build. 

Somehow I feel that I missed out something so do let me know if more information is needed in the documentation or any of the examples require change.

### CODE REVIEW

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -455,6 +470,19 @@ impl<T> From<Box<T>> for Pin<Box<T>> {
 
 #[stable(feature = "box_from_slice", since = "1.17.0")]
 impl<'a, T: Copy> From<&'a [T]> for Box<[T]> {
+    /// Converts a `&[T]` into a `Box<[T]>`
+    ///
+    /// This conversion does not allocate on the heap
```
Comment (sinkuu) [inline src/liballoc/boxed.rs]: This one (and therefore `From<&str>`) allocates a `RawVec`.

**Comment by TimNN** [comment]
Ping from triage @daniellimws: It looks like some changes are necessary to this PR.

**Comment by alexcrichton** [comment]
The documentation examples are tested and I think one of the examples adder here failed compilation (output in the full logs)

**Comment by daniellimws** [comment]
Really sorry about the hold up.

**Comment by TimNN** [comment]
Ping from triage @daniellimws: It looks like your PR failed on travis because one of the added examples fails to compile. (If you don't want to rely on travis / run the tests locally, you should be able to test the individual examples on [play.rust-lang.org](https://play.rust-lang.org/)).

**Comment by TimNN** [comment]
Ping from triage @daniellimws: Do you think you'll be able to update your PR?

**Comment by Centril** [comment]
Tests passed so I think we can land this for now.

@bors r+ rollup

**Pull request merged**